### PR TITLE
XOL-3558 Fetch balance_transaction when creating charge on Stripe

### PR DESCRIPTION
**Pull request by srishaharidas**

### CODE REVIEW

File: src/Util/StripeQueryAggregator.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+use \Guzzle\Http\QueryAggregator\QueryAggregatorInterface;
+use \Guzzle\Http\QueryString;
+
+class StripeQueryAggregator implements QueryAggregatorInterface
+{
+
+    /**
+     * Aggregates nested query string variables using [] but without the array index.
+     * Example: http://test.com?foo[]=1&foo[]=2
```
Comment (rushi) [inline src/Util/StripeQueryAggregator.php]: Give some better info. A before & after aggregation.

File: src/Message/AbstractRequest.php [hunk 2]
```diff
@@ -204,6 +277,49 @@ public function setSource($value)
...
+    protected function createClientRequest($data, array $headers = null)
+    {
+        // Stripe only accepts TLS >= v1.2, so make sure Curl is told
+        $config                          = $this->httpClient->getConfig();
+        $curlOptions                     = $config->get('curl.options');
+        $curlOptions[CURLOPT_SSLVERSION] = 6;
+        $config->set('curl.options', $curlOptions);
+        $this->httpClient->setConfig($config);
+
+        // Use custom query aggregator because Stripe only supports a specific format
```
Comment (rushi) [inline src/Message/AbstractRequest.php]: Explain _why_. It can be simply a reference to the documentation for `StripeQueryAggregator`

File: src/Message/PurchaseRequest.php [hunk 3]
```diff
@@ -109,4 +109,20 @@ private function validateLineItemAmounts(ItemBag $lineItems)
         }
         return $actualAmount == $this->getAmount();
     }
+
+    public function getEndpoint()
+    {
+        $endPoint = $this->endpoint . '/charges';
```
Comment (rushi) [inline src/Message/PurchaseRequest.php]: just do `parent::getEndPoint()`

**Review by rushi — CHANGES_REQUESTED** [review]

File: src/Util/StripeQueryAggregator.php (hunk 1 above)
Comment (srishaharidas) [reply src/Util/StripeQueryAggregator.php]: done

File: src/Message/AbstractRequest.php (hunk 2 above)
Comment (srishaharidas) [reply src/Message/AbstractRequest.php]: done

File: src/Message/PurchaseRequest.php (hunk 3 above)
Comment (srishaharidas) [reply src/Message/PurchaseRequest.php]: done

**Review by rushi — APPROVED** [review]

File: src/Message/RefundRequest.php [hunk 4]
```diff
@@ -128,17 +128,6 @@ public function getData()
...
-        if ($expandParams && is_array($expandParams)) {
-            $endPoint = $endPoint . '?';
-            foreach ($expandParams as $idx => $param) {
-                $endPoint .= "expand[]=" . urlencode($param);
-                if ($idx !== count($expandParams) - 1) {
-                    $endPoint .= "&";
-                }
-            }
-        }
-        return $endPoint;
```
Comment (anush) [inline src/Message/RefundRequest.php]: Why'd you revert this change?

Comment (srishaharidas) [reply src/Message/RefundRequest.php]: Was trying to do this on Saturday. This doen't actually work, a separate field needs to be expanded for refunds (`balance_transaction` of the `refunds` in the `charge`). Decided to do this in a separate PR since this was already in ready to deploy
